includes/class-wc-stripe-database-cache.php

```diff
@@ -0,0 +1,151 @@
+<?php
+
+defined( 'ABSPATH' ) || exit; // block direct access.
+
+/**
+ * Class WC_Stripe_Database_Cache
+ */
+
+/**
+ * A class for caching data as an option in the database.
+ */
+class WC_Stripe_Database_Cache {
+
+	/**
+	 * In-memory cache for the duration of a single request.
+	 *
+	 * This is used to avoid multiple database reads for the same data and as a backstop in case the database write fails,
+	 * thus ensuring the cache generator is not called multiple times (which would mean multiple API calls).
+	 *
+	 * @var array
+	 */
+	private static $in_memory_cache = [];
+
+	/**
+	 * Class constructor.
+	 */
+	private function __construct() {
+	}
+
+	/**
+	 * Stores a value in the cache.
+	 *
+	 * @param string $key  The key to store the value under.
+	 * @param mixed  $data The value to store.
+	 * @param int    $ttl  The TTL of the cache. Dafault 1 hour.
+	 *
+	 * @return void
+	 */
+	public static function set( $key, $data, $ttl = HOUR_IN_SECONDS ) {
+		self::write_to_cache( $key, $data, $ttl );
+	}
```

Comment:
I think we may need to require or add a key prefix so we don't unintentionally overwrite or delete "real" options in the database. Requiring callers to provide valid keys would make it more work for callers, while adding a prefix would reduce the key length that we can support, and could lead to redundant option key names. That said, having a key prefix would allow for clearing all persistent cached data via a database query.
```suggestion
	public static function set( $key, $data, $ttl = HOUR_IN_SECONDS ) {
		// Requiring a key prefix
		if ( ! self::is_permitted_key( $key ) ) {
		 	// Either return early and do nothing, or throw an exception
		 	return;
		 	throw new InvalidArgumentException( 'Unsupported key' );
		}
		// Add key prefix -- that function would simply do something like return 'wc_stripe_db_cache_' . $key;
		$db_key = self::get_db_key( $key );

		self::write_to_cache( $key, $data, $ttl );
	}
```

Reply:
To reduce the changes needed in the WC_Stripe_Payment_Method_Configurations class, I just kept the prefixes we had in place; I agree we should add a prefix for the cache (wcstripe_cache_*) but we should also handle the test/live mode internally, so we can remove the multiple WC_Stripe_Mode::is_test() ternary ifs we have everytime we use the cache. I'll make a new PR for that improvement.